ddprof-ffi/src/slice.rs

```diff
@@ -0,0 +1,259 @@
+use std::borrow::Cow;
+use std::fmt::{Debug, Formatter};
+use std::marker::PhantomData;
+use std::os::raw::c_char;
+use std::str::Utf8Error;
+
+/// Remember, the data inside of each member is potentially coming from FFI,
+/// so every operation
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct Slice<'a, T: 'a> {
+    ptr: *const T,
+    len: usize,
+    marker: PhantomData<&'a [T]>,
+}
+
+impl<'a, T: Debug> Debug for Slice<'a, T> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        unsafe { f.debug_list().entries(self.as_slice().iter()).finish() }
+    }
+}
+
+impl<'a, T: Eq> PartialEq<Self> for Slice<'a, T> {
+    fn eq(&self, other: &Self) -> bool {
+        unsafe {
+            return self.as_slice() == other.as_slice();
+        }
+    }
+}
+
+impl<'a, T: Eq> Eq for Slice<'a, T> {}
+
+// Use to represent strings -- should be valid UTF-8.
+pub type CharSlice<'a> = crate::Slice<'a, c_char>;
+
+/// Use to represent bytes -- does not need to be valid UTF-8.
+pub type ByteSlice<'a> = crate::Slice<'a, u8>;
+
+/// This exists as an intrinsic, but it is private.
+pub fn is_aligned_and_not_null<T>(ptr: *const T) -> bool {
+    !ptr.is_null() && ptr as usize % std::mem::align_of::<T>() == 0
+}
+
+impl<'a> Slice<'a, u8> {
+    /// # Safety
+    /// Slice needs to satisfy most of the requirements of std::slice::from_raw_parts but they can
+    /// relax the aligned and non-null requirements, as this function will detect these conditions
+    /// and return an empty slice instead.
+    pub unsafe fn as_bytes(&'a self) -> &'a [u8] {
+        if is_aligned_and_not_null(self.ptr) {
+            std::slice::from_raw_parts(self.ptr, self.len)
+        } else {
+            &[]
+        }
+    }
+
+    /// # Safety
+    /// This function has the same safety requirements as `as_bytes`.
+    pub unsafe fn try_to_utf8(&'a self) -> Result<&'a str, Utf8Error> {
+        std::str::from_utf8(self.as_bytes())
+    }
+
+    /// # Safety
+    /// This function has the same safety requirements as `as_bytes`
+    pub unsafe fn to_utf8_lossy(&'a self) -> Cow<'a, str> {
+        String::from_utf8_lossy(self.as_bytes())
+    }
+}
+
+impl<'a> Slice<'a, i8> {
```

Comment:
Both `i8` and `u8` will map to c_char, but on different platforms. Implementations are almost exactly the same. It could probably be made DRY with some trait, but I haven't fiddled with it and it didn't seem like too much code to repeat. For testing it's convenient for both `i8` and `u8` to be available, though, rather than just using `c_char`.